### PR TITLE
Update Datamodell 20151103 on delete

### PR DESCRIPTION
**Pull request by sjib**
ON Delete ergänzt

### CODE REVIEW

**Comment by sjib** [comment]
@m-kuhn Is this how you would like the ON DELETE 's?

**Comment by m-kuhn** [comment]
SQL-wise it looks good. Would be great to have some unit tests besides which check for correct behavior.

**Comment by sjib** [comment]
How would such a test look like?

e.g. Create a wastewater_structure object with cover and banquet and then delete it.

Could this be done with a create statement and then a delete statement?

How would you then check whether the child elements are also deleted?

 

Is there a simple check to check how many elements are in a table?

 

The same thing could be done for some other connected objects.

 

Would this be implemented in the the test.py file? Or an extra file?

 

Any ideas and comments from other members?

**Comment by m-kuhn** [comment]
On 11/04/2015 10:04 AM, Stefan wrote:
> How would such a test look like?
>
> e.g. Create a wastewater_structure object with cover and banquet and
> then delete it.
>
> Could this be done with a create statement and then a delete statement?
>
> How would you then check whether the child elements are also deleted?
> Is there a simple check to check how many elements are in a table?

That's what I would do, yes.

SELECT COUNT(*) FROM qgep.od_structure_part WHERE obj_id IN ('123', '456')

will return the number of features.

>
>
>
> The same thing could be done for some other connected objects.
>
>
>
> Would this be implemented in the the test.py file? Or an extra file?

A new test_relations.py or similar may be added.

>
>
>
> Any ideas and comments from other members?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/QGEP/datamodel/pull/6#issuecomment-153646407>.
>

-- 
Help to improve QGIS rendering possibilities before Nov 30, 2015
http://www.opengis.ch/2015/11/02/qgis-crowdfunding-2-5d-rendering

**Comment by sjib** [comment]
@m-kuhn How about this? test_relations.py

# Stefan Burckhardt

import unittest

import psycopg2
import psycopg2.extras
import decimal
from time import sleep

from utils import DbTestBase

class TestRelations(unittest.TestCase, DbTestBase):

    @classmethod
    def tearDownClass(cls):
        cls.conn.rollback()

    @classmethod
    def setUpClass(cls):
        cls.conn = psycopg2.connect("service=pg_qgep")


    def test_relations(self):
        # create cover element
        row = {
                'identifier': 'CO123',
                'level': decimal.Decimal('50.000')
        }

        obj_id = self.insert_check('vw_cover', row)
        
        row = {
                'identifier': 'CO456',
                'level': decimal.Decimal('60.656')
        }

        # create 2nd cover element
        obj_id2 = self.insert_check('vw_cover', row)

        # delete od_structure_part with obj_id
        row = self.select('od_structure_part', obj_id)

        cur = self.cursor()

        cur.execute("DELETE * FROM qgep.od_structure_part WHERE obj_id='{obj_id}'".format(obj_id=obj_id))
        
        # count amount of structure part elements and cover elements - should be one each
        amount_structure_part = cur.execute("SELECT COUNT(*) FROM qgep.od_structure_part WHERE obj_id IN ('CO123', 'CO456')")
        amount_cover = cur.execute("SELECT COUNT(*) FROM qgep.od_cover WHERE obj_id IN ('CO123', 'CO456')")
        
        
        assert amount_structure_part != amount_cover, "Relation test for structure_part - cover failed"


if __name__ == '__main__':
    unittest.main()

**Comment by sjib** [comment]
@m-kuhn Der Test macht soweit was ich will, ausser dass ich beim Count statement immer die Anzahl none bekomme. Damit kann ich auch nicht die beiden Zahlen vergleich. Was könnte da falsch sein?

FAIL: test_relations (test_relations.TestRelations)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/QGEP/datamodel/test/test_relations.py", line 66, in test_relations
    assert amount_structure_part != amount_cover, "Relation test for structure_part - cover failed"
AssertionError: Relation test for structure_part - cover failed
-------------------- >> begin captured stdout << ---------------------
obj_id = 00000000CO000000
obj_id2 = 00000000CO000001
Count after creation
amount_structure_part =  None
amount_cover = None
Count after deleting cover obj_id = 00000000CO000000
amount_structure_part =  None
amount_cover = None

**Comment by m-kuhn** [comment]
http://initd.org/psycopg/docs/cursor.html#cursor.execute

On 11/10/2015 03:55 PM, Stefan wrote:
>
> @m-kuhn <https://github.com/m-kuhn> Der Test macht soweit was ich
> will, ausser dass ich beim Count statement immer die Anzahl none
> bekomme. Damit kann ich auch nicht die beiden Zahlen vergleich. Was
> könnte da falsch sein?
>
>
>     FAIL: test_relations (test_relations.TestRelations)
>
> Traceback (most recent call last):
> File "/home/travis/build/QGEP/datamodel/test/test_relations.py", line
> 66, in test_relations
> assert amount_structure_part != amount_cover, "Relation test for
> structure_part - cover failed"
> AssertionError: Relation test for structure_part - cover failed
> -------------------- >> begin captured stdout << ---------------------
> obj_id = 00000000CO000000
> obj_id2 = 00000000CO000001
> Count after creation
> amount_structure_part = None
> amount_cover = None
> Count after deleting cover obj_id = 00000000CO000000
> amount_structure_part = None
> amount_cover = None
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/QGEP/datamodel/pull/6#issuecomment-155441438>.
>

-- 
Help to improve QGIS rendering possibilities before Nov 30, 2015
http://www.opengis.ch/2015/11/02/qgis-crowdfunding-2-5d-rendering

**Comment by sjib** [comment]
@m-kuhn Danke für den ersten Tipp. Jetzt sieht das Resultat eigentlich gut aus, aber das irgendwie stimmt hier der Vergleichsoperator noch nicht, wenn 
amount_structure_part =  [1L]
amount_cover = [1L]

assert amount_structure_part != amount_cover, "Relation test for structure_part - cover failed"

Muss ich noch Datentypen umwandeln? Oder das Element aus einer Liste auslesen? [1L] - sorry kenne mich noch zuwenig aus mit python datentypen.

FAIL: test_relations (test_relations.TestRelations)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/QGEP/datamodel/test/test_relations.py", line 70, in test_relations
    assert amount_structure_part != amount_cover, "Relation test for structure_part - cover failed"
AssertionError: Relation test for structure_part - cover failed
-------------------- >> begin captured stdout << ---------------------
obj_id = 00000000CO000000
obj_id2 = 00000000CO000001
Count after creation
amount_structure_part =  [2L]
amount_cover = [2L]
Count after deleting cover obj_id = 00000000CO000000
amount_structure_part =  [1L]
amount_cover = [1L]

File: test/test_relations.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        # count amount of structure part elements and cover elements - should be one each
+        # obj_id are 00000000CO000000 or 00000000CO000001
+        cur.execute("SELECT COUNT(*) FROM qgep.od_structure_part WHERE obj_id IN ('00000000CO000000', '00000000CO000001')")
+        amount_structure_part = cur.fetchone()
+        # amount_structure_part = cur.execute("SELECT COUNT(*) FROM qgep.od_structure_part WHERE identifier IN ('CO123', 'CO456')")
+
+        cur.execute("SELECT COUNT(*) FROM qgep.od_cover WHERE obj_id IN ('00000000CO000000', '00000000CO000001')")
+        amount_cover = cur.fetchone()
+        
+
```
Comment (m-kuhn) [inline test/test_relations.py]: Funktioniert es nicht, direkt die beiden Rückgabewerte zu vergleichen?

    amount_cover[0] == amount_structure_part[0]

**Comment by m-kuhn** [comment]
Superseded by #18

**Pull request closed without merging**